// @renproject/chains-ethereum/build/main/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumBaseChain = exports.EthereumConfigMap = void 0;
const interfaces_1 = require("@renproject/interfaces");
const utils_1 = require("@renproject/utils");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const web3_1 = __importDefault(require("web3"));
const networks_1 = require("./networks");
const utils_2 = require("./utils");
exports.EthereumConfigMap = {
    [interfaces_1.RenNetwork.Mainnet]: networks_1.renMainnet,
    [interfaces_1.RenNetwork.Testnet]: networks_1.renTestnet,
    [interfaces_1.RenNetwork.MainnetVDot3]: networks_1.renMainnetVDot3,
    [interfaces_1.RenNetwork.TestnetVDot3]: networks_1.renTestnetVDot3,
    [interfaces_1.RenNetwork.DevnetVDot3]: networks_1.renDevnetVDot3,
};
const isEthereumConfig = (renNetwork) => {
    return !!renNetwork.addresses;
};
const resolveNetwork = (renNetwork) => {
    if (!renNetwork) {
        return exports.EthereumConfigMap[interfaces_1.RenNetwork.Mainnet];
    }
    let networkConfig;
    if (renNetwork && isEthereumConfig(renNetwork)) {
        networkConfig = renNetwork;
    }
    else if (renNetwork) {
        const networkDetails = interfaces_1.getRenNetworkDetails(renNetwork);
        if (exports.EthereumConfigMap[networkDetails.name]) {
            networkConfig = exports.EthereumConfigMap[networkDetails.name];
        }
    }
    if (!networkConfig) {
        throw new Error(`Unrecognized network ${renNetwork}.`);
    }
    return networkConfig;
};
class EthereumBaseChain {
    constructor(web3Provider, renNetwork) {
        this.chain = EthereumBaseChain.chain;
        this.name = EthereumBaseChain.chain;
        this.legacyName = "Eth";
        this.logRequestLimit = undefined;
        this.utils = utils_1.utilsWithChainNetwork(EthereumBaseChain.utils, () => this.renNetworkDetails);
        this.getTokenContractAddress = async (asset) => {
            if (!this.web3 || !this.renNetworkDetails) {
                throw new Error(`${this.name} object not initialized - must provide network to constructor.`);
            }
            return utils_2.getTokenAddress(this.renNetworkDetails, this.web3, asset);
        };
        this.getGatewayContractAddress = async (token) => {
            if (!this.web3 || !this.renNetworkDetails) {
                throw new Error(`${this.name} object not initialized - must provide network to constructor.`);
            }
            const gatewayAddress = await utils_2.getGatewayAddress(this.renNetworkDetails, this.web3, token);
            if (gatewayAddress === "0x0000000000000000000000000000000000000000") {
                throw new Error(`Asset not supported on mint-chain.`);
            }
            return gatewayAddress;
        };
        this.withProvider = (web3Provider) => {
            this.web3 = new web3_1.default(web3Provider);
            return this;
        };
        /**
         * See [LockChain.initialize].
         */
        this.initialize = (renNetwork) => {
            this.renNetworkDetails =
                this.renNetworkDetails ||
                    exports.EthereumConfigMap[interfaces_1.getRenNetworkDetails(renNetwork).name];
            if (!this.renNetworkDetails) {
                throw new Error(`Unable to set ${this.name} network for RenVM network ${interfaces_1.getRenNetworkDetails(renNetwork).name}. Please provide ${this.name} network details to ${this.name} constructor.`);
            }
            return this;
        };
        // Supported assets
        this.assetIsNative = (asset) => {
            return asset === "ETH";
        };
        /**
         * `assetIsSupported` should return true if the asset is native to the
         * MintChain.
         *
         * ```ts
         * ethereum.assetIsSupported = asset => asset === "ETH";
         * ```
         */
        this.assetIsSupported = async (asset) => {
            if (this.assetIsNative(asset)) {
                return true;
            }
            if (!this.web3 || !this.renNetworkDetails) {
                throw new Error(`${this.name} object not initialized - must provide network to constructor.`);
            }
            // Check that there's a gateway contract for the asset.
            try {
                return !!(await this.getGatewayContractAddress(asset));
            }
            catch (error) {
                if (/(Empty address returned)|(Asset not supported on mint-chain)/.exec(String((error || {}).message))) {
                    // Ignore
                }
                else {
                    console.warn(error);
                }
                return false;
            }
        };
        /**
         * `assetDecimals` should return the number of decimals of the asset.
         *
         * If the asset is not supported, an error should be thrown.
         *
         
         */
        this.assetDecimals = (asset) => {
            if (asset === "ETH") {
                return 18;
            }
            throw new Error(`Unsupported asset ${asset}.`);
        };
        this.transactionID = (transaction) => {
            return transaction || "";
        };
        this.transactionFromID = (txid, _txindex) => utils_1.Ox(txid);
        this.transactionConfidence = async (transaction) => {
            if (!this.web3 || !this.renNetworkDetails) {
                throw new Error(`${this.name} object not initialized - must provide network to constructor.`);
            }
            if (transaction === null) {
                throw new Error(`Unable to fetch transaction confidence, transaction hash: ${transaction}`);
            }
            const currentBlock = new bignumber_js_1.default((await this.web3.eth.getBlockNumber()).toString());
            const receipt = await this.web3.eth.getTransactionReceipt(transaction);
            let current = 0;
            if (receipt.blockNumber) {
                const transactionBlock = new bignumber_js_1.default(receipt.blockNumber.toString());
                current = currentBlock.minus(transactionBlock).plus(1).toNumber();
            }
            return {
                current,
                target: this.renNetworkDetails.isTestnet ? 15 : 30,
            };
        };
        this.submitMint = async (asset, contractCalls, mintTx, eventEmitter) => {
            if (!mintTx.out) {
                throw new Error(`No signature passed to mint submission.`);
            }
            if (mintTx.out.revert !== undefined) {
                throw new Error(`Unable to submit reverted RenVM transaction.`);
            }
            if (!this.web3) {
                throw new Error(`${this.name} object not initialized - must provide network to constructor.`);
            }
            const existingTransaction = await this.findTransaction(asset, mintTx.out.nhash, mintTx.out.sighash);
            if (existingTransaction) {
                await utils_2.manualPromiEvent(this.web3, existingTransaction, eventEmitter);
                return existingTransaction;
            }
            return await utils_2.submitToEthereum(this.web3, contractCalls, mintTx, eventEmitter);
        };
        this.findTransaction = async (asset, nHash, sigHash) => {
            if (!this.renNetworkDetails || !this.web3) {
                throw new Error(`${this.name} object not initialized - must provide network to constructor.`);
            }
            return utils_2.findTransactionBySigHash(this.renNetworkDetails, this.web3, asset, nHash, sigHash, this.logRequestLimit);
        };
        this.resolveTokenGatewayContract = async (asset) => {
            if (!this.renNetworkDetails || !this.web3) {
                throw new Error(`${this.name} object not initialized - must provide network to constructor.`);
            }
            return utils_1.Ox(await utils_2.getTokenAddress(this.renNetworkDetails, this.web3, asset));
        };
        /**
         * Read a burn reference from an Ethereum transaction - or submit a
         * transaction first if the transaction details have been provided.
         */
        this.findBurnTransaction = async (asset, 
        // Once of the following should not be undefined.
        burn, eventEmitter, logger, timeout) => {
            if (!this.renNetworkDetails || !this.web3) {
                throw new Error(`${this.name} object not initialized - must provide network to constructor.`);
            }
            const { burnNonce, contractCalls } = burn;
            let { transaction } = burn;
            if (!transaction && burnNonce) {
                return utils_2.findBurnByNonce(this.renNetworkDetails, this.web3, asset, burnNonce.toString());
            }
            // There are three parameter configs:
            // Situation (1): A `burnNonce` is provided
            // Situation (2): Contract call details are provided
            // Situation (3): A transaction is provided
            // Handle situation (2)
            // Make a call to the provided contract and Pass on the
            // transaction hash.
            if (!transaction && contractCalls) {
                for (let i = 0; i < contractCalls.length; i++) {
                    const contractCall = contractCalls[i];
                    const last = i === contractCalls.length - 1;
                    const { contractParams, contractFn, sendTo } = contractCall;
                    const callParams = [
                        ...(contractParams || []).map((value) => value.value),
                    ];
                    const ABI = utils_1.payloadToABI(contractFn, contractParams);
                    const contract = new this.web3.eth.Contract(ABI, sendTo);
                    const txConfig = typeof contractCall === "object"
                        ? contractCall.txConfig
                        : {};
                    const config = await utils_2.withDefaultAccount(this.web3, Object.assign(Object.assign({}, txConfig), {
                        value: txConfig && txConfig.value
                            ? txConfig.value.toString()
                            : undefined,
                        gasPrice: txConfig && txConfig.gasPrice
                            ? txConfig.gasPrice.toString()
                            : undefined,
                    }));
                    logger.debug("Calling Ethereum contract", contractFn, sendTo, ...callParams, config);
                    const tx = contract.methods[contractFn](...callParams).send(config);
                    if (last) {
                        utils_2.forwardWeb3Events(tx, eventEmitter);
                    }
                    transaction = await new Promise((resolve, reject) => tx.on("transactionHash", resolve).catch((error) => {
                        try {
                            if (utils_2.ignorePromiEventError(error)) {
                                logger.error(utils_1.extractError(error));
                                return;
                            }
                        }
                        catch (_error) {
                            /* Ignore _error */
                        }
                        reject(error);
                    }));
                    logger.debug("Transaction hash", transaction);
                }
            }
            if (!transaction) {
                throw new Error(`Unable to find burn from provided parameters.`);
            }
            return utils_2.extractBurnDetails(this.web3, transaction, logger, timeout);
        };
        this.getFees = async (asset) => {
            if (!this.web3) {
                throw new Error(`${this.name} object not initialized - must provide network to constructor.`);
            }
            const gatewayAddress = await this.getGatewayContractAddress(asset);
            const mintFeeABI = {
                constant: true,
                inputs: [],
                name: "mintFee",
                outputs: [
                    {
                        internalType: "uint16",
                        name: "",
                        type: "uint16",
                    },
                ],
                payable: false,
                stateMutability: "view",
                type: "function",
            };
            const burnFeeABI = {
                constant: true,
                inputs: [],
                name: "burnFee",
                outputs: [
                    {
                        internalType: "uint16",
                        name: "",
                        type: "uint16",
                    },
                ],
                payable: false,
                stateMutability: "view",
                type: "function",
            };
            const gatewayContract = new this.web3.eth.Contract([mintFeeABI, burnFeeABI], gatewayAddress);
            const mintFee = await gatewayContract.methods.mintFee().call();
            const burnFee = await gatewayContract.methods.burnFee().call();
            return {
                mint: new bignumber_js_1.default(mintFee.toString()).toNumber(),
                burn: new bignumber_js_1.default(burnFee.toString()).toNumber(),
            };
        };
        this.getBalance = async (asset, address) => {
            const balanceOfABI = {
                constant: true,
                inputs: [
                    {
                        internalType: "address",
                        name: "account",
                        type: "address",
                    },
                ],
                name: "balanceOf",
                outputs: [
                    {
                        internalType: "uint256",
                        name: "",
                        type: "uint256",
                    },
                ],
                payable: false,
                stateMutability: "view",
                type: "function",
            };
            if (!this.web3) {
                throw new Error(`${this.name} object not initialized - must provide network to constructor.`);
            }
            const tokenAddress = await this.getTokenContractAddress(asset);
            const tokenContract = new this.web3.eth.Contract([balanceOfABI], tokenAddress);
            const balanceRaw = await await tokenContract.methods
                .balanceOf(address)
                .call();
            return new bignumber_js_1.default(balanceRaw.toString());
        };
        this.transactionRPCFormat = (transaction, _v2) => {
            utils_1.assertType("string | null", { transaction });
            if (transaction === null) {
                throw new Error(`Unable to encode transaction, transaction hash: ${transaction}`);
            }
            return {
                txid: utils_1.fromHex(transaction),
                txindex: "0",
            };
        };
        this.web3 = new web3_1.default(web3Provider);
        if (renNetwork) {
            this.renNetworkDetails = resolveNetwork(renNetwork);
        }
    }
}
exports.EthereumBaseChain = EthereumBaseChain;
EthereumBaseChain.chain = "Ethereum";
EthereumBaseChain.utils = {
    resolveChainNetwork: resolveNetwork,
    addressIsValid: utils_2.addressIsValid,
    addressExplorerLink: (address, network) => `${(EthereumBaseChain.utils.resolveChainNetwork(network) ||
        networks_1.renMainnet).etherscan}/address/${address}`,
    transactionExplorerLink: (transaction, network) => `${(EthereumBaseChain.utils.resolveChainNetwork(network) ||
        networks_1.renMainnet).etherscan}/tx/${transaction}`,
};
const _ = EthereumBaseChain;
