// @renproject/interfaces/build/main/networks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenNetworks = exports.getRenNetworkDetails = exports.RenNetwork = void 0;
var RenNetwork;
(function (RenNetwork) {
    RenNetwork["Mainnet"] = "mainnet";
    RenNetwork["Testnet"] = "testnet";
    // Staging
    RenNetwork["MainnetVDot3"] = "mainnet-v0.3";
    RenNetwork["TestnetVDot3"] = "testnet-v0.3";
    RenNetwork["DevnetVDot3"] = "devnet-v0.3";
})(RenNetwork = exports.RenNetwork || (exports.RenNetwork = {}));
const renMainnet = {
    name: RenNetwork.Mainnet,
    lightnode: "https://lightnode-mainnet.herokuapp.com",
    isTestnet: false,
};
const renTestnet = {
    name: RenNetwork.Testnet,
    lightnode: "https://lightnode-testnet.herokuapp.com",
    isTestnet: true,
};
const renMainnetVDot3 = {
    name: RenNetwork.MainnetVDot3,
    lightnode: "https://lightnode-new-mainnet.herokuapp.com/",
    isTestnet: false,
};
const renTestnetVDot3 = {
    name: RenNetwork.TestnetVDot3,
    lightnode: "https://lightnode-new-testnet.herokuapp.com/",
    isTestnet: true,
};
const renDevnetVDot3 = {
    name: RenNetwork.DevnetVDot3,
    lightnode: "https://lightnode-devnet.herokuapp.com/",
    isTestnet: true,
};
const getRenNetworkDetails = (renNetwork) => {
    switch (renNetwork) {
        case RenNetwork.Mainnet:
        case RenNetwork.Mainnet:
            return renMainnet;
        case RenNetwork.Testnet:
        case RenNetwork.Testnet:
            return renTestnet;
        case RenNetwork.MainnetVDot3:
        case RenNetwork.MainnetVDot3:
            return renMainnetVDot3;
        case RenNetwork.TestnetVDot3:
        case RenNetwork.TestnetVDot3:
            return renTestnetVDot3;
        case RenNetwork.DevnetVDot3:
        case RenNetwork.DevnetVDot3:
            return renDevnetVDot3;
        default:
            return renNetwork;
    }
};
exports.getRenNetworkDetails = getRenNetworkDetails;
exports.RenNetworks = [
    RenNetwork.Mainnet,
    RenNetwork.Testnet,
    RenNetwork.MainnetVDot3,
    RenNetwork.TestnetVDot3,
    RenNetwork.DevnetVDot3,
];
