// @renproject/chains-ethereum/build/module/polygon.js
import { getRenNetworkDetails, RenNetwork, } from "@renproject/interfaces";
import { Callable, utilsWithChainNetwork } from "@renproject/utils";
import { EthereumClass } from "./ethereum";
import { addressIsValid } from "./utils";
export const renPolygonTestnet = {
    name: "Polygon Testnet",
    chain: "polygonTestnet",
    isTestnet: true,
    chainLabel: "Polygon Testnet",
    networkID: 80001,
    infura: "https://rpc-mumbai.maticvigil.com/",
    etherscan: "https://testnet.ftmscan.com",
    addresses: {
        GatewayRegistry: "0xD881213F5ABF783d93220e6bD3Cc21706A8dc1fC",
        BasicAdapter: "0xD087b0540e172553c12DEEeCDEf3dFD21Ec02066",
    },
};
export const renPolygonMainnet = {
    name: "Polygon Mainnet",
    chain: "polygonMainnet",
    isTestnet: false,
    chainLabel: "Polygon Mainnet",
    networkID: 137,
    infura: "https://rpc-mainnet.maticvigil.com/",
    etherscan: "https://explorer.matic.network",
    addresses: {
        GatewayRegistry: "0x21C482f153D0317fe85C60bE1F7fa079019fcEbD",
        BasicAdapter: "0xAC23817f7E9Ec7EB6B7889BDd2b50e04a44470c5",
    },
};
export const PolygonConfigMap = {
    [RenNetwork.TestnetVDot3]: renPolygonTestnet,
    [RenNetwork.MainnetVDot3]: renPolygonMainnet,
};
const resolvePolygonNetwork = (renNetwork) => {
    if (!renNetwork) {
        return PolygonConfigMap[RenNetwork.MainnetVDot3];
    }
    if (renNetwork.addresses) {
        return renNetwork;
    }
    else {
        const details = getRenNetworkDetails(renNetwork);
        return details.isTestnet ? renPolygonTestnet : renPolygonMainnet;
    }
};
export class PolygonClass extends EthereumClass {
    constructor(web3Provider, renNetwork) {
        // To be compatible with the Ethereum chain class, the first parameter
        // is a web3Provider and the second the RenVM network. However,
        super(web3Provider, resolvePolygonNetwork(renNetwork));
        this.chain = PolygonClass.chain;
        this.name = PolygonClass.chain;
        this.legacyName = undefined;
        this.logRequestLimit = 1000;
        this.utils = utilsWithChainNetwork(PolygonClass.utils, () => this.renNetworkDetails);
        this.initialize = (renNetwork) => {
            this.renNetworkDetails =
                this.renNetworkDetails ||
                    PolygonConfigMap[getRenNetworkDetails(renNetwork).name];
            if (!this.renNetworkDetails) {
                throw new Error(`Unable to set ${this.name} network for RenVM network ${getRenNetworkDetails(renNetwork).name}. Please provide ${this.name} network details to ${this.name} constructor.`);
            }
            return this;
        };
    }
}
PolygonClass.chain = "Polygon";
PolygonClass.utils = {
    resolveChainNetwork: resolvePolygonNetwork,
    addressIsValid,
    addressExplorerLink: (address, network) => `${(Polygon.utils.resolveChainNetwork(network) ||
        renPolygonMainnet).etherscan}/address/${address}`,
    transactionExplorerLink: (transaction, network) => `${(Polygon.utils.resolveChainNetwork(network) ||
        renPolygonMainnet).etherscan}/tx/${transaction}`,
};
// @dev Removes any static fields, except `utils`.
export const Polygon = Callable(PolygonClass);
