// @renproject/utils/build/main/abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.payloadToMintABI = exports.payloadToABI = void 0;
const assert_1 = require("./assert");
const mintABITemplate = {
    constant: false,
    inputs: [
        {
            name: "_amount",
            type: "uint256",
        },
        {
            name: "_nHash",
            type: "bytes32",
        },
        {
            name: "_sig",
            type: "bytes",
        },
    ],
    outputs: [],
    payable: true,
    stateMutability: "payable",
    type: "function",
};
const payloadToABI = (methodName, payload) => {
    // Type validation
    assert_1.assertType("string", { methodName });
    (payload || []).map(({ type, name }) => assert_1.assertType("string", { type, name }));
    return [
        {
            name: methodName,
            type: "function",
            inputs: [
                ...(payload || []).map((value) => ({
                    type: value.type,
                    name: value.name,
                })),
            ],
            outputs: [],
        },
    ];
};
exports.payloadToABI = payloadToABI;
const payloadToMintABI = (methodName, payload) => {
    // Type validation
    assert_1.assertType("string", { methodName });
    (payload || []).map(({ type, name }) => assert_1.assertType("string", { type, name }));
    return [
        Object.assign(Object.assign({}, mintABITemplate), { name: methodName, inputs: [
                ...(payload || []).map((value) => ({
                    type: value.type,
                    name: value.name,
                })),
                ...(mintABITemplate.inputs ? mintABITemplate.inputs : []),
            ] }),
    ];
};
exports.payloadToMintABI = payloadToMintABI;
